File: compress/press.cpp
#include "generator.h"
#include "mystl.h"
#include <iostream>
#include <cassert>

using mystl::istream;
using std::ostream;
using mystl::cin;
using std::cout;
using std::cerr;
using std::fill;
using std::div;

static Generator<unsigned> codes(istream &is)
{
    unsigned n = 0;
    bool flag = false;

    for (int c; (c = is.get()) != -1;)
    {
        if (isdigit(c))
        {
            flag = true;
            n = n * 10 + c - '0';
        }
        else
        {
            if (flag)
                co_yield n;

            flag = false;
            n = 0;
        }
    }
}

int main()
{
    ostream *os = &cout;
    os->put(0x1f);
    os->put(0x9d);
    os->put(0x90);
    unsigned cnt = 0, nbits = 9;
    const unsigned bitdepth = 16;
    char buf[20] = {0};
    
    for (auto codes = ::codes(cin); codes;)
    {
        unsigned code = codes();
        unsigned *window = (unsigned *)(buf + nbits * (cnt % 8) / 8);
        *window |= code << (cnt % 8) * (nbits - 8) % 8;
        ++cnt;

        if (cnt % 8 == 0 || code == 256)
        {
            os->write(buf, nbits);
            fill(buf, buf + sizeof(buf), 0);
        }

        if (code == 256)
            nbits = 9, cnt = 0;
        
        if (nbits != bitdepth && cnt == 1U << nbits - 1)
            ++nbits, cnt = 0;
    }

    auto dv = div((cnt % 8) * nbits, 8);
    os->write(buf, dv.quot + (dv.rem ? 1 : 0));
    os->flush();
    return 0;
}

File: compress/press.py
#!/usr/bin/pypy3

import sys

if __name__ == "__main__":
    out = sys.stdout.buffer
    out.write(b'\x1f\x9d\x90')
    nbits = 9
    cnt = n = 0
    for lin in sys.stdin:
        c = int(lin)
        n = c << nbits * (cnt % 8) | n
        cnt += 1
        if cnt % 8 == 0 or c == 256:
            out.write(n.to_bytes(nbits, 'little'))
            n = 0
        if c == 256:
            nbits = 9
            cnt = 0
        if cnt == 1 << nbits - 1 and nbits != 16:
            nbits += 1
            cnt = 0
    a = divmod(nbits * (cnt % 8), 8)
    out.write(n.to_bytes(a[0] + (1 if a[1] else 0), 'little'))


